include/ebpf_api.h

```diff
@@ -417,6 +417,35 @@ extern "C"
     _Must_inspect_result_ ebpf_result_t
     ebpf_get_program_info_from_verifier(_Outptr_ const ebpf_program_info_t** program_info);
 
+    typedef struct _ebpf_test_run_options
+    {
+        _Readable_bytes_(data_size_in) const uint8_t* data_in; ///< Input data to the program.
+        _Writable_bytes_(data_size_out) uint8_t* data_out;     ///< Output data from the program.
+        size_t data_size_in;                                   ///< Size of input data.
+        size_t data_size_out; ///< Maximum length of data_out on input and actual length of data_out on output.
+        _Readable_bytes_(context_size_in) const uint8_t* context_in; ///< Input context to the program.
+        _Writable_bytes_(context_size_out) uint8_t* context_out;     ///< Output context from the program.
+        size_t context_size_in;                                      ///< Size of input context.
+        size_t context_size_out; ///< Maximum length of context_out on input and actual length of context_out on output.
```

Comment:
Can these two lines ever be different?  There are helpers to resize data, but to my knowledge the context size should never change.

Reply:
That's how the libbpf API is defined. Asking on slack if there is a reason for this.

Reply:
This is used to support cases where the application doesn't know the size of the context structure (i.e. it might change across OS versions). Should we remove this as we aren't tied to OS releases?

Reply:
I don't follow.  If you don't know the size of the context structure, what do you pass in context_size_in?   Is it assumed that you can pass a higher value in than will actually be used?

Reply:
The idea is that you can pass a context buffer that is smaller than the current context. The remaining fields in the context structure are then zero initialized. 

As per @dylandreimerink:

> Yes. If you are testing a socket program for example which takes a struct __sk_buff  ctx, the userspace might not be up-to-date with changes made. If in the newest kernel a field has been added at the end of struct __sk_buff  then you might provide an input ctx of lets say 50 bytes and the BPF program will execute with the last field set to zero, which is often supported for this kind of forward/backward compatibility.
If the linux kernel detects that ctx_size_out is not 0 and set to a value smaller than the CTX the kernel has, it will set the appropriate size and return a -ENOSPC error code.